sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs

```diff
@@ -0,0 +1,240 @@
+        }
+
+        private async Task Download(bool async, CancellationToken cancellationToken)
+        {
+            Response<T> response;
+
+            HttpRange range = new HttpRange(_position, _bufferSize);
+
+            try
+            {
+                response = async
+                    ? await _downloadToAsyncFunc(range, _requestConditions, default, cancellationToken).ConfigureAwait(false)
+                    : _downloadToFunc(range, _requestConditions, default, cancellationToken);
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex);
+                throw;
+            }
+
+            Stream networkStream = (Stream)typeof(T).GetProperty("Content").GetValue(response.Value, null);
+
+            _stream.SetLength(0);
+
+            if (async)
+            {
+                await networkStream.CopyToAsync(
+                    _stream,
+                    Constants.DefaultBufferSize,
+                    cancellationToken)
+                    .ConfigureAwait(false);
+            }
+            else
+            {
+                networkStream.CopyTo(
+                    _stream,
+                    Constants.DefaultBufferSize);
+            }
+
+            _stream.Position = 0;
+            _lastDownloadBytes = response.GetRawResponse().Headers.ContentLength.GetValueOrDefault();
+            _length = GetBlobLength(response);
+        }
+
+        private static void ValidateReadParameters(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException($"{nameof(buffer)}", $"{nameof(buffer)} cannot be null.");
+            }
+
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException($"{nameof(offset)} cannot be less than 0.");
+            }
+
+            if (offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException($"{nameof(offset)} cannot exceed {nameof(buffer)} length.");
+            }
+
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException($"{nameof(count)} cannot be less than 0.");
+            }
+
+            if (offset + count > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException($"{nameof(offset)} + {nameof(count)} cannot exceed {nameof(buffer)} length.");
+            }
+        }
+
+        private static long GetBlobLength(Response<T> response)
+        {
+            response.GetRawResponse().Headers.TryGetValue("Content-Range", out string lengthString);
+            string[] split = lengthString.Split('/');
+            return Convert.ToInt64(split[1], CultureInfo.InvariantCulture);
+        }
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => false;
+
+        public override long Length => _length;
```

Comment:
Concerned about someone getting a stream and us telling them it's zero length when we actually don't know.

Reply:
I'll fix this.  I'm thinking since we have `OpenRead` and `OpenReadAsync`, we can fetch the blob size when we are initializing the `LazyLoadingBlobStream`.